godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,638 @@
+    use crate::obj::Gd;
+    use crate::sys::{
+        GDExtensionBool, GDExtensionCallError, GDExtensionConstStringNamePtr,
+        GDExtensionConstVariantPtr, GDExtensionInt, GDExtensionMethodInfo, GDExtensionObjectPtr,
+        GDExtensionPropertyInfo, GDExtensionScriptInstanceDataPtr, GDExtensionScriptLanguagePtr,
+        GDExtensionStringPtr, GDExtensionTypePtr, GDExtensionVariantPtr, GDExtensionVariantType,
+        GodotFfi, PtrcallType, GDEXTENSION_CALL_OK,
+    };
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    fn borrow_instance_mut<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> RefMut<'a, T> {
+        unsafe { (*instance).inner.borrow_mut() }
+    }
+
+    fn borrow_instance<'a, T: ScriptInstance>(instance: *mut ScriptInstanceData<T>) -> Ref<'a, T> {
+        unsafe { (*instance).inner.borrow() }
+    }
+
+    fn try_borrow_instance<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> Result<Ref<'a, T>, BorrowError> {
+        unsafe { (*instance).inner.try_borrow() }
+    }
+
+    /// # Safety
+    /// - we expect the engine to provide a vallid instance pointer
+    fn instance_data_as_script_instance<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+    ) -> *mut ScriptInstanceData<T> {
+        p_instance as *mut ScriptInstanceData<T>
+    }
+
+    /// # Safety
+    /// - we expect the engine to provide a valid const string name pointer
+    unsafe fn transfer_string_name_from_godot(p_name: GDExtensionConstStringNamePtr) -> StringName {
+        StringName::from_arg_ptr(p_name as GDExtensionTypePtr, PtrcallType::Standard)
+    }
+
+    fn transfer_bool_to_godot(value: bool) -> u8 {
+        value as u8
+    }
+
+    /// # Safety
+    /// - we expect the engine to provide a valid variant pointer the return value can be moved into
+    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: GDExtensionVariantPtr) {
+        variant.move_return_ptr(
+            return_ptr as GDExtensionTypePtr,
+            godot_ffi::PtrcallType::Standard,
+        )
+    }
+
+    /// # Safety
+    /// - the engine has to provide a valid `*mut u32` pointer for the list length
+    /// - the list pointer is guaranteed to point to a list that has an equal length and capacity
+    unsafe fn transfer_ptr_list_to_godot<T>(ptr_list: Vec<T>, list_length: *mut u32) -> *const T {
+        *list_length = ptr_list.len() as u32;
+
+        let ptr = Box::into_raw(ptr_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    fn transfer_script_to_godot(script: Gd<crate::engine::Script>) -> GDExtensionObjectPtr {
+        let script = mem::ManuallyDrop::new(script.into_ffi());
+
+        script.sys() as GDExtensionObjectPtr
+    }
+
+    /// # Safety
+    /// - ptr is expect to point to a list with both a length and capacity of list_length.
+    /// - it's expected that the list pointer has been created with `transfer_ptr_list_to_godot`
```

Comment:
```suggestion
    /// # Safety
    /// - `ptr` is expected to point to a list with both a length and capacity of `list_length`.
    /// - The list pointer `ptr` must have been created with `transfer_ptr_list_to_godot`.
```
(I think the 2nd assumption implies the 1st, no?)